File: packages/ogdf.js/src/layouts/energybased/nr.js
import { PARAMETER_TYPE } from '../../utils/parameters'
import createLayout from '../creater'

const NAME = 'NR'
export const OUTER_PARAMETER_DEFINITION = {
    useWorker: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: false
    }
}

export const ORIGIN_PARAMETER_DEFINITION = {
    bendNormalizationAngle: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [-Infinity, Infinity],
        default: Math.PI
    },
    desiredMinEdgeLength: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 20.0
    },
    dummyInsertionThreshold: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 5
    },
    gravitation: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 1.0 / 16.0
    },
    initDummiesPerEdge: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 1
    },
    initialTemperature: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 10.0
    },
    maxDisturbance: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 0
    },
    maxDummiesPerEdge: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 3
    },
    minDistCC: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 30.0
    },
    minimalTemperature: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 1.0
    },
    numberOfIterations: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 30000
    },
    oscillationAngle: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: Math.PI / 2.0
    },
    pageRatio: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 1.0
    },
    postProcessing: {
        type: PARAMETER_TYPE.CATEGORICAL,
        range: ['None', 'KeepMultiEdgeBends', 'Complete'],
        default: 'Complete'
    },
    randomInitialPlacement: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: true
    },
    repulsionDistance: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 40.0
    },
    temperatureDecreaseOffset: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, 1],
        default: 0.0
    }
}

const nr = createLayout(NAME, OUTER_PARAMETER_DEFINITION, ORIGIN_PARAMETER_DEFINITION, {})
export default nr

File: packages/ogdf.js/src/entry/energybased/nr.cpp
#include <ogdf/energybased/NodeRespecterLayout.h>
#include "../main.h"

EM_PORT_API(NodeRespecterLayout *)
LayoutModule_NodeRespecterLayout(double bendNormalizationAngle, double desiredMinEdgeLength, double dummyInsertionThreshold, double gravitation, int initDummiesPerEdge, double initialTemperature, double maxDisturbance, int maxDummiesPerEdge, double minDistCC, double minimalTemperature, int numberOfIterations, double oscillationAngle, double pageRatio, int postProcessing, bool randomInitialPlacement, double repulsionDistance, double temperatureDecreaseOffset)
{
    //LayoutModule
    NodeRespecterLayout *model = new NodeRespecterLayout();

    model->setBendNormalizationAngle(bendNormalizationAngle);
    model->setDesiredMinEdgeLength(desiredMinEdgeLength);
    model->setDummyInsertionThreshold(dummyInsertionThreshold);
    model->setGravitation(gravitation);
    model->setInitDummiesPerEdge(initDummiesPerEdge);
    model->setInitialTemperature(initialTemperature);
    model->setMaxDisturbance(maxDisturbance);
    model->setMaxDummiesPerEdge(maxDummiesPerEdge);
    model->setMinDistCC(minDistCC);
    model->setMinimalTemperature(minimalTemperature);
    model->setNumberOfIterations(numberOfIterations);
    model->setOscillationAngle(oscillationAngle);
    model->setPageRatio(pageRatio);
    model->setPostProcessing(static_cast<NodeRespecterLayout::PostProcessingMode>(postProcessing));
    model->setRandomInitialPlacement(randomInitialPlacement);
    model->setRepulsionDistance(repulsionDistance);
    model->setTemperatureDecreaseOffset(temperatureDecreaseOffset);

    return model;
}

EM_PORT_API(float *)
NR(int node_num, int link_num, int *source, int *target, double bendNormalizationAngle, double desiredMinEdgeLength, double dummyInsertionThreshold, double gravitation, int initDummiesPerEdge, double initialTemperature, double maxDisturbance, int maxDummiesPerEdge, double minDistCC, double minimalTemperature, int numberOfIterations, double oscillationAngle, double pageRatio, int postProcessing, bool randomInitialPlacement, double repulsionDistance, double temperatureDecreaseOffset)
{
    node *nodes;
    Graph G;
    GraphAttributes GA(G, GraphAttributes::nodeGraphics | GraphAttributes::edgeGraphics);

    nodes = new node[node_num];
    for (int i = 0; i < node_num; i++)
    {
        nodes[i] = G.newNode();
    }

    edge e;

    for (int i = 0; i < link_num; i++)
    {
        e = G.newEdge(nodes[source[i]], nodes[target[i]]);
        GA.bends(e);
    }
    //LayoutModule
    NodeRespecterLayout *model = new NodeRespecterLayout();

    model->setBendNormalizationAngle(bendNormalizationAngle);
    model->setDesiredMinEdgeLength(desiredMinEdgeLength);
    model->setDummyInsertionThreshold(dummyInsertionThreshold);
    model->setGravitation(gravitation);
    model->setInitDummiesPerEdge(initDummiesPerEdge);
    model->setInitialTemperature(initialTemperature);
    model->setMaxDisturbance(maxDisturbance);
    model->setMaxDummiesPerEdge(maxDummiesPerEdge);
    model->setMinDistCC(minDistCC);
    model->setMinimalTemperature(minimalTemperature);
    model->setNumberOfIterations(numberOfIterations);
    model->setOscillationAngle(oscillationAngle);
    model->setPageRatio(pageRatio);
    model->setPostProcessing(static_cast<NodeRespecterLayout::PostProcessingMode>(postProcessing));
    model->setRandomInitialPlacement(randomInitialPlacement);
    model->setRepulsionDistance(repulsionDistance);
    model->setTemperatureDecreaseOffset(temperatureDecreaseOffset);

    model->call(GA);

    float *re = (float *)malloc(node_num * 2 * 4);
    for (int i = 0; i < node_num; ++i)
    {
        re[i * 2] = GA.x(nodes[i]);
        re[i * 2 + 1] = GA.y(nodes[i]);
    }

    return re;
}